communication/src/coap_channel_new.cpp

```diff
@@ -0,0 +1,1457 @@
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    bool getBlock = false;
+    if (size > 0) {
+        if (msg->pos == msg->end) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        if (curMsgId_ != msg->id) {
+            // TODO: Support asynchronous reading from multiple message instances
+            LOG(ERROR, "CoAP message buffer is no longer available");
+            return SYSTEM_ERROR_NOT_SUPPORTED;
+        }
+        auto bytesToRead = std::min<size_t>(size, msg->end - msg->pos);
+        if (data) {
+            std::memcpy(data, msg->pos, bytesToRead);
+        }
+        msg->pos += bytesToRead;
+        if (msg->pos == msg->end) {
+            releaseMessageBuffer();
+            if (msg->hasMore.value_or(false)) {
+                if (blockCallback) {
+                    assert(msg->type == MessageType::RESPONSE); // TODO: Support cloud-to-device blockwise requests
+                    auto resp = staticPtrCast<ResponseMessage>(msg);
+                    // Send a new request with the original options and updated block number
+                    auto req = resp->blockRequest;
+                    assert(req && req->blockIndex.has_value());
+                    ++req->blockIndex.value();
+                    CHECK(prepareMessage(req));
+                    CHECK(sendMessage(std::move(req)));
+                    resp->state = MessageState::WAIT_BLOCK;
+                    addRefToList(blockResps_, std::move(resp));
+                    getBlock = true;
+                } else {
+                    LOG(WARN, "Incomplete read of blockwise response");
+                }
+            }
+        }
+        size = bytesToRead;
+    }
+    msg->blockCallback = blockCallback;
+    msg->errorCallback = errorCallback;
+    msg->callbackArg = callbackArg;
+    return getBlock ? COAP_RESULT_WAIT_BLOCK : 0;
+}
+
+int CoapChannel::peekPayload(coap_message* apiMsg, char* data, size_t size) {
+    auto msg = RefCountPtr(reinterpret_cast<CoapMessage*>(apiMsg));
+    if (msg->sessionId != sessId_) {
+        return SYSTEM_ERROR_COAP_REQUEST_CANCELLED;
+    }
+    if (msg->state != MessageState::READ) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    if (size > 0) {
+        if (msg->pos == msg->end) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        if (curMsgId_ != msg->id) {
+            // TODO: Support asynchronous reading from multiple message instances
+            LOG(ERROR, "CoAP message buffer is no longer available");
+            return SYSTEM_ERROR_NOT_SUPPORTED;
+        }
+        size = std::min<size_t>(size, msg->end - msg->pos);
+        if (data) {
+            std::memcpy(data, msg->pos, size);
+        }
+    }
+    return size;
+}
+
+void CoapChannel::destroyMessage(coap_message* apiMsg) {
+    if (!apiMsg) {
+        return;
+    }
+    auto msg = RefCountPtr<CoapMessage>::wrap(reinterpret_cast<CoapMessage*>(apiMsg)); // Take ownership
+    clearMessage(msg);
```

Comment:
This is the pattern I would expect in `removeRefFromList`. First take ownership of the reference counted object, remove it from the list, then let the smart pointer go out of scope to decrement the reference count.

Reply:
Discussed [here](https://github.com/particle-iot/device-os/pull/2715#discussion_r1468230265).